### ISSUE TITLE
Form schema doesn't work with structs, because they don't register in container

### ISSUE THREAD

**Issue by Kukunin** [dry-rb/dry-validation#263]
I use latest `dry-validation` 0.10.3 with separated `dry-structs`.

Let me demonstrate the bug with code:

``` ruby
Types = Dry::Types.module
class A < Dry::Struct
  attribute :var, Types::Strict::FilledString
end

# works fine
S = Dry::Validation.Schema do
  required(:a).value(type?: A)
end

# gives Dry::Container::Error: Nothing registered with the key "a"
F = Dry::Validation.Form do
  required(:a).value(type?: A)
end
```

It doesn't make difference between `filled` or `value` method

This is stacktrace for your convenience: 

``` ruby
Dry::Container::Error: Nothing registered with the key "a"
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-container-0.5.0/lib/dry/container/resolver.rb:22:in `block in call'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-container-0.5.0/lib/dry/container/resolver.rb:21:in `fetch'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-container-0.5.0/lib/dry/container/resolver.rb:21:in `call'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-container-0.5.0/lib/dry/container/mixin.rb:108:in `resolve'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-container-0.5.0/lib/dry/container/mixin.rb:121:in `[]'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types.rb:65:in `block in []'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/concurrent-ruby-1.0.2/lib/concurrent/map.rb:133:in `block in fetch_or_store'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/concurrent-ruby-1.0.2/lib/concurrent/map.rb:122:in `fetch'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/concurrent-ruby-1.0.2/lib/concurrent/map.rb:132:in `fetch_or_store'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types.rb:56:in `[]'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types/compiler.rb:30:in `visit_type'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types/compiler.rb:15:in `visit'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types/compiler.rb:75:in `visit_key'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types/compiler.rb:15:in `visit'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types/compiler.rb:80:in `block in merge_with'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/dry-types-0.9.0/lib/dry/types/compiler.rb:80:in `map'
... 13 levels...
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands/console.rb:9:in `start'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands/commands_tasks.rb:68:in `console'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands.rb:17:in `<top (required)>'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:274:in `require'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:274:in `block in require'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:240:in `load_dependency'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:274:in `require'
    from /Users/kukunin/Sites/matic-www/bin/rails:8:in `<top (required)>'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:268:in `load'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:268:in `block in load'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:240:in `load_dependency'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:268:in `load'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    from /Users/kukunin/.rbenv/versions/2.3.0/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    from -e:1
```

I understand, that struct support is experimental, so I don't expect hotfix for this problem. Just want let you know about it. 

Also, I'm going to migrate schemas to use `type_specs`. Hopefully it will resolve this problem.
If you have other idea for other workaround, let me know.

Thanks

**Comment by Kukunin** [comment]
```ruby
F = Dry::Validation.Form do
  configure { config.type_specs = true }
  required(:a, A).value(type?: A)
end
```

works fine

**Comment by solnic** [comment]
`type?` should work with any class, regardless if it's a struct or not, so I'm marking this as a bug and I'll try to fix it in 0.10.4

**Comment by flash-gordon** [comment]
This will be fixed in the next release, works fine in dry-schema

```ruby
M = Class.new
S = Dry::Schema.Form do
  required(:a).value(type?: M)
end

2.6.1 :001 > S.(a: 1)
 => #<Dry::Schema::Result{:a=>1} errors={:a=>["must be M"]}>
2.6.1 :001 > S.(a: M.new)
 => #<Dry::Schema::Result{:a=>#<M:0x00007f8c739a0e28>} errors={}>
```
